### ISSUE TITLE
the speed of the model

### ISSUE THREAD

**Issue by 18222278799** [MaartenGr/BERTopic#468]
For unsupervised data, running a dataset is slow. Are there any running time statistics for different sizes? Is there any way to improve the running speed?

**Comment by MaartenGr** [comment]
Can you give me a bit more information? Are you using a GPU? How large is your data? Which version of BERTopic are you using?
Also, could you share the code that you are using to train your model?

**Comment by 18222278799** [comment]
Thank you. I use about 50,0000 sentences dataset. using GPU but only run 1.1GB/8GB.   version is 0.9.3
 I use the code: 
vectorizer_model = CountVectorizer(ngram_range=(1, 1), stop_words="english", min_df=1) // when I set min_df=100, errors:max_df corresponds to < documents than min_df, I also dont know the reason.
dc = df['content'].apply(lambda x: re.split('https:\/\/.*', str(x))[0]).apply(lambda x: removestop(x))
a = dc.tolist()

topic_model = BERTopic(vectorizer_model=vectorizer_model,calculate_probabilities=False,verbose=True)
topics, probs = topic_model.fit_transform(a)
topic_model.reduce_topics(a, topics, nr_topics=20)

**Comment by MaartenGr** [comment]
> I use about 50,0000 sentences dataset.

Just to be clear, are we talking about 50,000 sentences or 500,000 sentences? 

> For unsupervised data, running a dataset is slow.

When you mention that training on your dataset is slow, how long is it exactly running that makes you feel like it is slow? 

> using GPU but only run 1.1GB/8GB

Could you check which GPU you are using and if it is used following [this](https://stackoverflow.com/a/48152675/10532563) StackOverflow post?

> a = dc.tolist()

Lastly, could you check that `a = dc.tolist()` indeed contains 50?

Another thing that you can try is installing the newest version of HDBSCAN with `pip install --upgrade hdbscan`. The 0.8.28 version solves some issue with large data which might speed up your code.

**Comment by 18222278799** [comment]
Sorry.  Yeah, it's 500,000 sentences.
I run it about 6 hours. I dont know whether it's reasonable.
yes, I make sure that it uses 1GB GPU
the a is a list which contains 500,000 sentences. I dont know whether it's true when using. Because all sentences in one twitter datasets, i'd like to know the top 20 topics of the whole dataset

**Comment by MaartenGr** [comment]
Hmmm, the 6 hours can be long but that highly depends on your GPU. Which GPU are you using? Also, I saw that you use `verbose=True`. Typically, it outputs how long each step has taken. Can you share that output?

**Comment by 18222278799** [comment]
I use 2070, which has 12GB, but the task only use 1.1GB, I dont know the reason. I think the final hdbsan process is so slow.

And I have another question which is more important than speed. After I got the topics, I find many topics have many same words, which is very difficult to distinguish.
Like:

[('london', 0.016165334963697042), ('lockdown', 0.013676706817772884), ('beer', 0.012433005111945613), ('nhs', 0.011442232869942648), ('coronavirus', 0.011191893338268855), ('uk', 0.010263366432955306), ('brewers', 0.008800705067169302), ('tweet', 0.008327613572779667), ('support', 0.008144877845572647), ('pandemic', 0.007981718976888675)]
[('coronavirus', 0.013294063988809392), ('pandemic', 0.011996244141107545), ('support', 0.011894729393009295), ('nhs', 0.009826260906432122), ('health', 0.008716449925394033), ('care', 0.008091610836093959), ('crisis', 0.008066650611246359), ('workers', 0.006581720863005995), ('vulnerable', 0.006554186863985224), ('staff', 0.0065408430035483025)]
[('food', 0.012616259233449898), ('lockdown', 0.012411261239560623), ('work', 0.012374008779056946), ('coronavirus', 0.012215146851516692), ('support', 0.009366778388559328), ('nhs', 0.00865680273501425), ('health', 0.00819349249160839), ('pandemic', 0.007511451362701211), ('working', 0.007375135142436187), ('workers', 0.006575690160866968)]
[('vaccine', 0.013169435435967271), ('nhs', 0.012606994000411055), ('coronavirus', 0.012541068079575347), ('pandemic', 0.008752242645297496), ('staff', 0.00796236993958997), ('london', 0.007588281692483863), ('health', 0.00737453325641647), ('patients', 0.007172430325412421), ('vaccines', 0.006766791454600087), ('flu', 0.006625298400539063)]
[('government', 0.014482371638584103), ('capitalism', 0.011285565092466811), ('pandemic', 0.009662735937571857), ('coronavirus', 0.009097414109594035), ('nhs', 0.008872981404211697), ('deaths', 0.008653981287921083), ('health', 0.007575049553106031), ('vaccine', 0.007269221262680239), ('crisis', 0.006950487551289056), ('support', 0.0068156607513834315)]
[('schools', 0.012073011936226787), ('coronavirus', 0.010464849947221265), ('children', 0.009391686214851185), ('london', 0.008239402712509426), ('health', 0.008206010599102472), ('support', 0.007876962301804142), ('school', 0.007829678676112358), ('pandemic', 0.006693903900703064), ('vaccine', 0.00664060768690282), ('virus', 0.00659892273455744)]
[('lockdown', 0.014824097776755531), ('coronavirus', 0.014645859336632113), ('london', 0.012276776648946447), ('nhs', 0.011641612092439631), ('liverpool', 0.00987604653645618), ('manchester', 0.007507341374966966), ('united', 0.00748852523263343), ('england', 0.007387150698121569), ('leeds', 0.007287990087657286), ('pandemic', 0.0071727206226038925)]
[('vaccine', 0.02882964439201315), ('coronavirus', 0.024179698549122185), ('vaccinated', 0.014651200212625101), ('vaccines', 0.012968243509933997), ('vaccination', 0.012502268020524704), ('coviduk', 0.010224896301840938), ('pandemic', 0.009031760274323196), ('covidvaccine', 0.008643182735241929), ('coronavirusuk', 0.007649983510034987), ('getvaccinated', 0.007546833884009336)]
[('health', 0.016654539370411336), ('nhs', 0.01479001935394174), ('care', 0.013242516530498025), ('patients', 0.012759674198025905), ('support', 0.011321940650474297), ('pandemic', 0.010556178714438175), ('coronavirus', 0.009675440163839623), ('nurses', 0.0096202627202605), ('mentalhealth', 0.009273711516453809), ('vaccine', 0.00719787189088617)]
[('music', 0.015028450342209482), ('coronavirus', 0.012957188016291189), ('pandemic', 0.010043038460674256), ('borisjohnson', 0.00905769098928545), ('vaccine', 0.008396980785942195), ('london', 0.008138102601065433), ('government', 0.007891104249274054), ('nhs', 0.007779409240643615), ('boris', 0.006919632932574092), ('coviduk', 0.006406199433201136)]
[('christmas', 0.021004536461449605), ('coronavirus', 0.01756428908399), ('petition', 0.016758273861007218), ('lockdown', 0.016498038529180904), ('governments', 0.013676062760558999), ('pandemic', 0.013177558480307194), ('public', 0.011444024660995266), ('london', 0.010781395219397605), ('government', 0.008377223425849295), ('nhs', 0.007997319907502187)]
[('la', 0.041017552486569214), ('en', 0.039800831628065754), ('el', 0.03474305630382689), ('es', 0.014393207634655372), ('los', 0.014081044059249138), ('coronavirus', 0.01217889554557555), ('una', 0.012052217074761392), ('london', 0.009805373693411105), ('le', 0.009648887472538665), ('est', 0.008159939981164422)]
[('mohammad', 0.026727826445969723), ('nasser', 0.02576094797308182), ('allah', 0.02497281819583225), ('coronavirus', 0.01363181382166729), ('mahdi', 0.013330160247870102), ('pandemic', 0.00952601513293232), ('ramadan', 0.008773004262319311), ('muslim', 0.008116336406953913), ('health', 0.006760511153586402), ('thailand', 0.006650443980043764)]
[('lockdown', 0.019404652239603817), ('mask', 0.017578107477691108), ('coronavirus', 0.013701328653289636), ('masks', 0.013388770220791435), ('face', 0.011574705841031633), ('kittiesinlockdown', 0.010946480619939193), ('london', 0.009203687563001406), ('wear', 0.008233927334429009), ('mainecoonkittens', 0.007830333397805369), ('pandemic', 0.007092119160682558)]

**Comment by MaartenGr** [comment]
> I use 2070, which has 12GB, but the task only use 1.1GB, I dont know the reason. I think the final hdbsan process is so slow.

It indeed can be that HDBSCAN is relatively slow but I cannot be sure without the logging that results from setting `verbose=True`. Moreover, HDBSCAN and UMAP are not GPU accelerated which means that a GPU is only helpful for creating the embeddings. 

> Is this because I use topic_model.reduce_topics(a, topics, nr_topics=20)?

It definitely could be. If you have 500,000 documents and the you keep the `min_topic_size` at 15, then it is likely that hundreds if not thousands of topics will be created. If you are then reducing those topics into only 20 topics, it becomes difficult to separate them cleanly. In practice, I would advise setting `min_topic_size` at around 200 and seeing how many topics are generated. Then, you can reduce it to 20 topics in order to keep some separation. 

However, it is unlikely that 500,000 sentences only contain 20 topics. Instead, I would advise not reducing the number of topics to 20 and just keeping the top 20 most frequent topics instead. I think that they key here is to use a higher `min_topic_size` (e.g., 200) to get a reasonable amount of topics (e.g., 100) and then take the top 20 most frequent topics.
